### PR TITLE
Video Optimization: Improve handling of files exceeding upload limit

### LINKED ISSUE #6580: Video Optimization: Handling of files exceeding upload limit
## Feature Description

Scenario: the site's upload limit is 100MB. You upload a file that is 150MB large.

Currently, since the file exceeds the limit, it is rejected for upload.

However, it _might_ be possible that after optimization the file is only 90MB large. (Note: we don't know that beforehand).

In https://github.com/google/web-stories-wp/pull/6541#discussion_r585463733 we discussed that we could try to attempt transcoding to see if it works.

Idea: the site's upload limit is 100MB. You upload a file that is 150MB large.

The file is being transcoded (could take a minute or so) and after that, the size is checked again.

If file is now smaller than 100MB -> upload
If file is still larger than  100MB -> tell user that file could not be uploaded despite transcoding.

Pros:

* We could allow uploading files that are originally larger but get smaller due to transcoding

Cons:

* File could still be larger after transcoding and thus fail, which seems confusing for users 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Scenario: the site's upload limit is 100MB. You upload a file that is 150MB large.

Original size: 170MB (mp4) Transcode size: 6 MB

## Testing Instructions

1. Download a large video. [Example](https://www.pexels.com/video/different-kinds-of-vehicles-on-the-freeway-2053100/).
2. Create a new story.
3. Drag video into canvas. 
4. Video should transcode and upload. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6580

### CODE REVIEW

**Comment by spacedmonkey** [comment]
FYI, most of this change is whitespace changes.

File: assets/src/edit-story/app/uploader/useUploader.js [hunk 1]
```diff
@@ -82,41 +82,41 @@ function useUploader() {
...
+              /* translators: %s: list of allowed file types. */
+              __('Please choose only %s to upload.', 'web-stories'),
+              translateToExclusiveList(allowedFileTypes)
+            );
+          }
+
+          throw createError('ValidError', file.name, message);
+        }
+        // TODO: Move this check to useUploadMedia?
+      } else if (isFileTooLarge) {
```
Comment (swissspidy) [inline assets/src/edit-story/app/uploader/useUploader.js:119]: I don't think `else if` is correct here.

This `isFileTooLarge` check is for `MEDIA_TRANSCODING_MAX_FILE_SIZE`, i.e. to see if the file is too large for transcoding.

Let's add some more unit tests for `useUploader` so we know there's no regression.

Comment (spacedmonkey) [reply assets/src/edit-story/app/uploader/useUploader.js:119]: > I don't think else if is correct here.

Why not?

To be honest, I reviewed the existing tests and they seemed pretty solid. What kind of tests should add?

Comment (swissspidy) [reply assets/src/edit-story/app/uploader/useUploader.js:119]: This introduction of new code without breaking any existing tests proves that there's no test coverage for this newly changed logic.

Comment (spacedmonkey) [reply assets/src/edit-story/app/uploader/useUploader.js:119]: > This introduction of new code without breaking any existing tests proves that there's no test coverage for this newly changed logic.

Maybe. Either way I have added some new tests.

File: assets/src/edit-story/app/uploader/test/useUploader.js [hunk 2]
```diff
@@ -141,6 +141,38 @@ describe('useUploader', () => {
...
+    });
+
+    it('throws an error if file too large to transcode or upload', async () => {
+      const {
+        actions: { validateFileForUpload },
+      } = setup({});
+
+      await expect(() =>
+        validateFileForUpload(
+          { size: 2147483649, type: 'video/mp4' },
```
Comment (swissspidy) [inline assets/src/edit-story/app/uploader/test/useUploader.js]: Nit: other tests use `1024 * 1024 * 1024 * 2` for improved readability to grasp more quickly what size it represents.

File: assets/src/edit-story/app/uploader/test/useUploader.js [hunk 3]
```diff
@@ -141,6 +141,38 @@ describe('useUploader', () => {
       );
     });
 
+    it('throws an error if file too large to transcode', async () => {
```
Comment (swissspidy) [inline assets/src/edit-story/app/uploader/test/useUploader.js:144]: This case is already covered by the identical `throws an error if file is too large for transcoding` test below.

What's missing is a case for `validateFileForUpload( { size: 1024 * 1024 * 150, type: 'video/mp4' }, true, false )` ("file is too large to upload but can be transcoded", the case this PR changes) and `validateFileForUpload( { size: 1024 * 1024 * 50, type: 'video/mp4' }, false, false )` ("file is not too large to upload").

Jus tinkering with this locally, the test cases for all scenarios would look like this:

```js
it('throws an error if file is too large for transcoding', async () => {
  const {
    actions: { validateFileForUpload },
  } = setup({
    maxUpload: 1024 * 1024 * 1024 * 10,
  });

  await expect(() =>
    validateFileForUpload(
      {
        size: 1024 * 1024 * 1024 * 3,
        type: 'video/quicktime',
      },
      true,
      true
    )
  ).toThrow(
    'Your file is too large (3072 MB) and cannot be processed. Please try again with a file that is smaller than 2048 MB.'
  );
});

it('does not throw an error if file is too large to upload but can be transcoded', async () => {
  const {
    actions: { validateFileForUpload },
  } = setup({});

  await expect(() =>
    validateFileForUpload(
      { size: 1024 * 1024 * 150, type: 'video/mp4' },
      true,
      false
    )
  ).not.toThrow();
});

it('throws an error if file is too large to transcode or upload', async () => {
  const {
    actions: { validateFileForUpload },
  } = setup({});

  await expect(() =>
    validateFileForUpload(
      { size: 1024 * 1024 * 1024 * 2, type: 'video/mp4' },
      false,
      true
    )
  ).toThrow(
    'Your file is 2048MB and the upload limit is 100MB. Please resize and try again!'
  );
});

it('does not throw an error if file is not too large to upload', async () => {
  const {
    actions: { validateFileForUpload },
  } = setup({});

  await expect(() =>
    validateFileForUpload(
      { size: 1024 * 1024 * 50, type: 'video/mp4' },
      false,
      false
    )
  ).not.toThrow();
});
```

**Review by swissspidy — APPROVED** [review]

**Pull request merged**